### PR TITLE
Add context to source and sink

### PR DESCRIPTION
**Pull request by jerrypeng**
### Motivation

Add context to source and sink.

Allows users to record custom metrics in custom sources and sinks

### CODE REVIEW

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/ContextImpl.java [hunk 1]
```diff
@@ -56,7 +58,7 @@
 /**
  * This class implements the Context interface exposed to the user.
  */
-class ContextImpl implements Context {
+class ContextImpl implements Context, SinkContext, SourceContext {
```
Comment (rdhabalia) [inline pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/ContextImpl.java]: does it make sense to have a base interface as `Context` and other contexts can implements it
such as: `SinkContext`, `SourceContext`, `FunctionContext`?

**Comment by rdhabalia** [comment]
is it possible to add unit test-cases for the changes?

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/ContextImpl.java (hunk 1 above)
Comment (jerrypeng) [reply pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/ContextImpl.java]: well Context is the "function" context and we can't really change that now our we will break backwards compatibility. 

The APIs SinkContext, SourceContext, and Context is and will be different.  The three won't have the all the same APIs but currently we are only exposing one method in SinkContext and SourceContext thus we can just use ContextImpl but in the future that might change

**Comment by jerrypeng** [comment]
@rdhabalia I have added unit tests

**Comment by jerrypeng** [comment]
retest this please

**Comment by jerrypeng** [comment]
retest this please

**Comment by jerrypeng** [comment]
retest this please

**Pull request merged**